### PR TITLE
[DEVOPS-834] Cardano faucet package

### PR DESCRIPTION
**Pull request by boothead**
## Description

Faucet for the cardano testnet.

## Linked issue

[DEVOPS-834](https://iohk.myjetbrains.com/youtrack/issue/DEVOPS-834)

## Type of change

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [ ] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: default.nix [hunk 1]
```diff
@@ -92,10 +93,30 @@ let
       mkDerivation = args: super.mkDerivation (args // {
         enableLibraryProfiling = enableProfiling;
         enableExecutableProfiling = enableProfiling;
-        # Static linking for everything to work around
-        # https://ghc.haskell.org/trac/ghc/ticket/14444
-        # This will be the default in nixpkgs since
-        # https://github.com/NixOS/nixpkgs/issues/29011
```
Comment (cleverca22) [inline default.nix]: leave the comments in, and move `enableSharedExecutables` out of the `args ? src` block

File: faucet/default.nix [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+, cardano-sl-core, cardano-sl-crypto, cardano-sl-wallet
+, cardano-sl-wallet-new, connection, cryptonite, data-default, ekg
+, ekg-core, ekg-statsd, exceptions, http-client, http-client-tls
+, lens, log-warper, memory, mmorph, mtl, optparse-applicative
+, QuickCheck, serokell-util, servant, servant-client
+, servant-client-core, servant-server, servant-swagger
+, servant-swagger-ui, stdenv, swagger2, text, text-format, tls, wai
+, wai-cors, wai-extra, warp
+}:
+mkDerivation {
```
Comment (cleverca22) [inline faucet/default.nix]: can probably delete this file, stack2nix handles it automatically

File: faucet/generated-wallet-details.json [hunk 3]
```diff
@@ -0,0 +1 @@
+{"wallet-id":"Ae2tdPwUPEZJkLTchA9mETtyXumBT1K1gqtX3dgXUguJFLuvyNU9BzQXrEK","recovery-words":["relax","secret","seek","story","supreme","meat","judge","onion","weasel","sponsor","garlic","erase"],"address":"DdzFFzCqrhswmBoqDubnsZZpvnfFbjgwLwFyk4vC5jJJf75m36f4M3f67k4iia1RQdCUPmJHV7c9TK2ogfzjoQSYHKCoM4YTv6zrmr28","account-index":2147483648}
```
Comment (erikd) [inline faucet/generated-wallet-details.json]: Is it worth pretty printing that JSON so its a bit more human readable?

File: faucet/server/Main.hs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        ecfg <- eitherDecode <$> BSL.readFile cfgFile
+        either (error . ("Error decoding: " ++)) return ecfg
+      _ -> error "Need a --config argument pointing to a json file"
+    runLogger config $ do
+      fEnv <- initEnv config (serverMetricStore ekg)
+      let server = hoistServer faucetDocAPI (nat fEnv) faucetHandler
+      _statsd <- liftIO $ forkStatsd (config ^. fcStatsdOpts . _Wrapped') (fEnv ^. feStore)
+      liftIO $ run (config ^. fcPort) (serve faucetDocAPI server)
+  where
+    -- runLogger :: LoggerNameBox IO a -> IO a
```
Comment (erikd) [inline faucet/server/Main.hs]: Comment? Why not actually have the type sig there so the compiler can validate that it is indeed correct.

File: faucet/generated-wallet-details.json (hunk 3 above)
Comment (boothead) [reply faucet/generated-wallet-details.json]: That wasn't actually supposed to be in there. I've added it to ignore.

**Comment by boothead** [comment]
@rvl these would be easier to address if they were inline :-)

> Split up the withdraw function.

I've split out the queue handling parts 59a71c7bb 

> Can we have at least one token test case?

I'm sure there are some parts that can be sensibly tested now

> Needs a little documentation.

There are haddocks on all the main functions and swagger docs (under `/docs`). What form would you like the documentation in? Do you mean a better README?

> Remove _dead code and commented out code

I've removed all the commented code I could find, could you review inline anymore you see and anything you think is dead? 7cabdb9c7

> Remove nix files and .projectile.

Done a0158671a

> Remove index.html

This is kinda handy to have around for testing purposes... Now that recaptcha is enabled you can't really test with `curl` or a rest client

> Copyright IOHK

Done - found the remaining instance in LICENSE 3d945c61c 

> There is a missing letter in Withdrawl

Urban dictionary informs me that "withdrawl" is

    > What a Southerner suffers when he moves to the North

which isn't quite what I intended. Fixed in 1c2ae37cc

**Comment by boothead** [comment]
Made a start on a test suite in 67cd2dfeb

File: faucet/cardano-sl-faucet.cabal [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+                    , Cardano.Faucet.Types.Config
+                    , Cardano.Faucet.Types.API
+                    , Cardano.Faucet.Types.Recaptcha
+                    , Cardano.Faucet.Types
+                    , Cardano.Faucet.Metrics
+                    , Cardano.Faucet.Init
+                    , Cardano.WalletClient
+  default-language: Haskell2010
+
+executable faucet
```
Comment (rvl) [inline faucet/cardano-sl-faucet.cabal]: It would help with deployment if this were called `cardano-faucet`.

Comment (boothead) [reply faucet/cardano-sl-faucet.cabal]: Sure - I can change that easy enough.

File: faucet/server/Main.hs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import           System.Remote.Monitoring (forkServer, serverMetricStore)
+import           System.Remote.Monitoring.Statsd (forkStatsd)
+import           System.Wlog (LoggerNameBox, launchFromFile, usingLoggerName)
+
+import           Cardano.Faucet
+import           Cardano.Faucet.Swagger
+
+main :: IO ()
+main = withCompileInfo $ do
+    ekg <- forkServer "localhost" 8001
```
Comment (rvl) [inline faucet/server/Main.hs]: This needs to be configurable with `--ekg-server` option, like the wallet.
You might be able to re-use option handling code from `Pos.Infra.Statistics.Ekg`.

**Comment by rvl** [comment]
Thanks. Can `readCaptchaSecret` read just the first line of the file and not include the newline characters?

**Comment by boothead** [comment]
> existing-wallet-details.json -- does this file need to be present?

Good catch - removed

> Cardano.Faucet.Endpoints.withdraw function

> (optional) split out captcha verification handling into a separate function.

Didn't really see any benefit of moving the catcha handling somewhere else, so I've left it.

> the use of whenJust with prisms seems obtuse. why not just a pattern matching function?
Cardano.Faucet.Init

Agreed, switched to pattern match. 

> why is this module called Init? `Cardano.Faucet.Init.createWallet`

Because it only exports one function which initializes the `FaucetEnv`

> Cardano.Faucet.Init.processWithdrawals
> LambdaCase could make it shorter

By one line, didn't really seem worth it..

> Can be split up into little functions with descriptive names.

Again - not sure I see the benefit here. `processWithdrawals` is only 25 lines or so, 8 of which is logging?

> the code which posts the transaction should be factored out from the code which gets a payment off the queue and puts the result.

Don't really see the point - it's just another place to jump to in the code if you need to understand what's going on..?

Cardano.Faucet.Metrics

decrWithDrawn isn't used -- can be removed.
Cardano.Faucet.Types.API

Yup, removed.

> instance FromJSON WithdrawalRequest -- can it be made to handle a missing "g-recaptcha-response" by providing the empty string as a default value.

I'm not sure I understand the logic here. I could change the recaptcha response to a `Maybe` I suppose and allow that in the case that the secret isn't set in the config?

> DepositRequest/Result -- if it's not currently used then remove it. The same goes for any other dead code.

Removed.

> formatting library can be used with logging to avoid having show and pack everywhere.

True - doesn't really seem worth the effort to change now though TBH. Happy to if you want though.

> There are some modules with -Wall added. Is it possible to enable that for all modules or too inconvenient?

I've enabled -Wall in the cabal file

**Comment by rvl** [comment]
Ported faucet branch from `develop` to `release/1.3.0`.

Previous version of faucet which is based off develop is on the [devops-834-faucet-develop](https://github.com/input-output-hk/cardano-sl/tree/devops-834-faucet-develop) branch.

**Pull request merged**